### DISCUSSION TITLE
Export Full Canvas When Zoomed In

### DISCUSSION THREAD

**Discussion by chonigman** [fabricjs/fabric.js#6963] [Q&A] [answered]
So I have a limited canvas size that does not necessarily match the viewport (i.e. the canvas is larger than the viewport but still limited to a max size). I'm trying to figure out how to use toDataURL and export the full canvas and it's contents even when you are zoomed in. Is there a way to reset zoom and export without rendering it? Or is there another way to achieve this? What I'm going for is similar to a minimap but I really just need a static image that shows full size of canvas despite what the user is currently viewing that lives outside the main canvas.

Thanks!

**Comment by asturur** [comment]
did you check the minimap tutorial? is incomplete but has sone useful code.
Is on fabricjs.com

**Reply by chonigman** [reply]
I have, but I'm pretty turning off Object Caching will negatively impact the app I am building. Is there no way to have the export detached from the viewport? I will try implementing what's in the minimap view, but I really want just a static image that gets updated sometimes and is non-interactive. Appreciate the help.

**Reply by asturur** [reply]
So i'll try to explain myself better.
Obejcts cached are drawn as in image and conserved till their size fit for rendering.
When you render the whole canvas in a small or big export, the objects get rendered at a different size, cache removed and recreated, then for the next canvas render, cache is killed again.

This may seems fast, but is creating a lot of new canvases.

What fabricJS is missing is an export functionality that would allow to render once in a while a picture without influencing the current canvas objects. I created one but is not compatible with clipPath and that is why i could not merge it.

You do not have to disable caching to follow the minimap tutorial.
You can export the full canvas to an image using the toDataUrl or toCanvasElement methods and asking to ignore the viewport.

**Reply by chonigman** [reply]
I was unable to figure out how to export the full canvas to an image using toDataUrl. I pass in arguments for the full width and height of the canvas at normal zoom, and then the left and top values are set to 0. This works fine if the viewport is not zoomed in at all, however, if zoom > 1, it exports the viewport with the zoom and not the entire canvas. In my app, the canvas is almost always larger than what is visible, but has a max size available for drawing. Users can pan and zoom around the larger space, so I want to have a small rendering of the whole canvas available to see. Currently I am making a copy of the canvas object and rendering the data url  from that which seems to work. I will have to go back and see if the minimap tutorial will suffice. 

Thanks!

**Reply by asturur** [reply]
This is a simple way to export the canvas at full size, then go back to where you were.

```js
  function createCanvasEl() {
    var originalVPT = fabricCanvas.viewportTransform;
    fabricCanvas.viewportTransform = [
      1, 0, 0, 1, 0, 0
    ];
    var canvas = fabricCanvas.toCanvasElement(1);
    fabricCanvas.viewportTransform = originalVPT;
    return canvas;
  }
```

**Reply by syarkhan** [reply]
As asturur mentioned, one way is to transform the viewport to original size and call toDataURL and then set the viewport to its previous state like this:

`let viewport = canvas.viewportTransform;
        canvas.viewportTransform = [1, 0, 0, 1, 0, 0];
        let dataUrl = canvas.toDataURL({
            multiplier: 1,
            format: "png",
            height: artboard.height,
            width: artboard.width
        });
        canvas.viewportTransform = viewport;` 

another way can be to insert a hidden <canvas> tag to html and when calling toDataURL, init a StaticCanvas by canvas id with width and height of artboard and use toJSON and loadFromJSON and then call toDataURL to export full canvas, after that you can call canvas.clear() and canvas.dispose()

**Comment by chonigman** [comment, marked as answer]
EDIT: (Please see comment below for real solution)

Thanks for the help y'all. I went down the route of having a static hidden canvas that I load the main canvas' json to and export the jpg from there. 

The complications I came up against was that there are views where the user's canvas size/aspect is different from the full size, this plus a variable zoom state made things a little weird. I had issues using the viewportTransform as it was changing the size of the canvas in their view.

**Reply by asturur** [reply]
it shouldn't change it unless something asyncronous is going on, and it is if you also use toJSON/ fromJSON.
I m nearly sure it should work. Are you able to reproduce your issue in a jsfiddle so that we can look at it?

**Reply by chonigman** [reply]
oof. I was doing it in an async function and at some point introduced toJSON to load into a separate canvas. Setting/resetting the viewport does seem to do the trick. Thanks for the assistance. 

For posterity this is the code that worked for. 

 ```
     let vpt = canvas.viewportTransform;
     canvas.viewportTransform = [1, 0, 0, 1, 0, 0];
     const data = canvas.toDataURL({
       width: fullWidth,
       height: height,
       multiplier: 1,
       left: 0,
       top: 0,
       format: "jpg",
     });
     canvas.viewportTransform = vpt;
```

**Reply by 752337625** [reply]
Can you share your minimap code
